File: app/javascript/shared/tables/officers.js
import escape from "lodash/escape";
import padStart from "lodash/padStart";
import {int_renderer, employee_id_renderer, employee_name_renderer, earnings_renderer, date_renderer, zip_renderer, text_renderer, title_renderer, unknown} from "renderers";
import {initDataTable} from "shared/tables/common";

initDataTable("table.officers-table", function($table, options) {
  $table.DataTable({
    ...options,
    columns: [
      {data: "employee_id", render: employee_id_renderer},
      {data: "badge", render: badge_renderer, orderable: false},
      {data: "name", render: employee_name_renderer},
      {data: "title", render: text_renderer},
      {data: "organization", render: title_renderer},
      {data: "doa", render: date_renderer},
      {data: "postal", render: zip_renderer},
      {data: "state", render: text_renderer},
      {data: "neighborhood", render: text_renderer, orderable: false},
      {data: "articles_officers_count", render: articles_officers_count_renderer},
      {data: "ia_score", render: ia_score_renderer},
      {data: "details_count", render: int_with_link_renderer("details")},
      {data: "field_contacts_count", render: int_with_link_renderer("field_contacts")},
      {data: "incidents_count", render: int_with_link_renderer("incidents")},
      {data: "citations_count", render: int_with_link_renderer("citations")},
      {data: "swats_count", render: int_with_link_renderer("swats")},
      {data: "total", render: earnings_renderer},
      {data: "regular", render: earnings_renderer},
      {data: "retro", render: earnings_renderer},
      {data: "other", render: earnings_renderer},
      {data: "overtime", render: earnings_renderer},
      {data: "injured", render: earnings_renderer},
      {data: "detail", render: earnings_renderer},
      {data: "quinn", render: earnings_renderer},
    ],
    order: [[16, 'desc']]
  });
});

function badge_renderer(data, type, row) {
  if (type != "display") {
    return data;
  }
  if (data == null) {
    return unknown();
  }
  if (/^\d+$/.test(data)) {
    data = padStart(data, 5, "0");
  }
  return data;
}

function ia_score_renderer(data, type, row) {
  if (type != "display") {
    return data;
  }
  if (data == null) {
    return `<div class="text-center">${unknown()}</div>`;
  }
  return `<div class="text-center"><a class="officer__ia-score-${data}" href="${escape(row['url'])}#complaints">${escape(data)}</a></div>`;
}

function articles_officers_count_renderer(data, type, row) {
  let ret =  `<a href="${escape(row['url'])}#articles">${escape(data)}</a>`;
  if (row.articles_officers_to_review_count) {
    ret = `${ret} <span class="text-muted"> / ${row.articles_officers_to_review_count}</span>`;
  }
  ret = `<div class="text-center">${ret}</div>`;
  return ret;
}

function int_with_link_renderer(fragment) {
  return (data, type, row) => {
    if (type != "display") {
      return data;
    }
    if (data == null) {
      return `<div class="text-center">${unknown()}</div>`;
    }
    return `<div class="text-center"><a href="${escape(row['url'])}#${fragment}">${escape(data)}</a></div>`;
  };
}

File: app/lib/exporter/officers.rb
class Exporter::Officers < Exporter::Exporter
  def column_definitions
    column("active") { write_boolean(record.active) }
    column("employee_id") { record.employee_id }
    column("name") { record.name }
    column("organization") { record.organization }
    column("title") { record.title }
    column("doa") { record.doa }
    column("badge") { record.badge }
    column("zip_code") { write_zip_code(record.postal) }
    column("city") { record.zip_code ? record.zip_code.city : nil }
    column("state") { record.zip_code ? record.zip_code.state : nil }
    column("neighborhood") { record.zip_code ? record.zip_code.neighborhood : nil }
    column("regular") { write_money(record.regular) }
    column("retro") { write_money(record.retro) }
    column("other") { write_money(record.other) }
    column("overtime") { write_money(record.overtime) }
    column("injured") { write_money(record.injured) }
    column("detail") { write_money(record.detail) }
    column("quinn") { write_money(record.quinn) }
    column("total") { write_money(record.total) }
    column("rank") { record.rank }
    column("ia_score") { record.ia_score }
    column("ia_sustained_conduct_unbecoming") { record.ia_sustained_conduct_unbecoming.count }
    column("ia_sustained_neg_duty") { record.ia_sustained_neg_duty.count }
    column("ia_sustained_respectful_treatment") { record.ia_sustained_respectful_treatment.count }
    column("ia_sustained_self_identification") { record.ia_sustained_self_identification.count }
    column("ia_sustained_use_of_force") { record.ia_sustained_use_of_force.count }
    column("ia_sustained_details") { record.ia_sustained_details.count }
    column("ia_sustained_cases") { record.ia_sustained_complaints.count }
    column("ia_sustained_allegations") { record.ia_sustained_allegations.count }
    column("ia_cases") { record.ia_complaints }
    column("ia_allegations") { record.ia_allegations }
    column("field_contacts_count") { record.field_contacts_count }
    column("incidents_count") { record.incidents_count }
    column("complaints_count") { record.complaints_count }
    column("swats_count") { record.swats_count }
    column("details_count") { record.details_count }
    column("citations_count") { record.citations_count }
    column("articles_officers_count") { record.articles_officers_count }
    column("url") { officer_url(record) }
  end

  def records
    Officer.includes(:zip_code, :complaint_officers, :complaints, complaint_officers: [:complaint])
  end
end
